File: app/javascript/channels/chat_room_channel.js
import consumer from "./consumer"

consumer.subscriptions.create({ channel: "ChatRoomChannel", room_id: "your_room_id" }, {
  connected() {
    // Called when the subscription is ready for use on the server
    console.log("Connected to chat room channel");
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
    console.log("Disconnected from chat room channel");
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
    $('#messages').append(data.message);
  },

  send_message(message) {
    this.perform('receive', { content: message });
  }
});

$(document).on('turbolinks:load', function() {
  $('#message_form').on('submit', function(e) {
    e.preventDefault();
    const message = $('#message_input').val();
    consumer.subscriptions.subscriptions[0].send_message(message);
    $('#message_input').val('');
  });
});

File: app/channels/chat_room_channel.rb
class ChatRoomChannel < ApplicationCable::Channel
  def subscribed
    # stream_from "some_channel"
    stream_from "chat_room_#{params[:room_id]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def send_message(data)
    message = Message.create!(content: data['content'], user: current_user)
    ActionCable.server.broadcast(
      "chat_room_#{params[:room_id]}", message: render_message(message),
                                       user: message.user.username
    )
  end

  private

  def render_message(message)
    ApplicationController.render(partial: 'messages/message', locals: { message: })
  end
end
